README
Libraries required in addition to those obtained by auvlibs.sh:
   lib3ds-dev tcl-dev tk-dev


gts-0.7.6/ChangeLog
Fri Mar 24 11:47:26 NZST 2006  Stephane Popinet <s.popinet@niwa.co.nz>
  * Changed 'gtstransform' to 'transform' in Debian package

Fri Mar 10 12:48:43 NZDT 2006  Stephane Popinet <s.popinet@niwa.co.nz>
  * Marcello's debian patch for ARM and Alpha was breaking floating-point control!

Fri Mar 10 10:41:16 NZDT 2006  Stephane Popinet <s.popinet@niwa.co.nz>
  * Debian snapshot packages were not correctly versioned

Sun Mar  5 21:11:09 NZDT 2006  Stephane Popinet <popinet@users.sf.net>
  * Removed 'arch' junk

Fri Feb  3 23:50:14 NZDT 2006  Stephane Popinet <popinet@users.sf.net>
  * Updated macros for automake 1.8

Thu Jan 26 06:48:18 NZDT 2006  Stephane Popinet <popinet@users.sf.net>
  * predicates_init.h has nothing to do in darcs

Thu Jan 26 06:45:30 NZDT 2006  Stephane Popinet <popinet@users.sf.net>
  * LC_ALL is set to POSIX in all executables
  
  This fixes problems when running the tools/examples on non-English systems.

Mon Dec  5 10:34:17 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Changelog is manual for debian package

Fri Dec  2 09:58:57 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Debian packages use correct versioning

Thu Dec  1 18:11:53 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Updated macro docs for new version of gtk-doc

Thu Dec  1 17:12:38 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Floating-point control for Alpha and ARM (from M. Magallon debian port)

Thu Dec  1 17:07:48 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Debian snapshot creation (based on Marcelo Magalon official debian packages)

Tue Nov 29 15:44:36 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fixed doc for gts_triangle_quality()

Tue Nov 29 13:17:28 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Removed USE_ROBUST_PREDICATES option (predicates are always robust now)

Mon Oct 17 11:21:35 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  tagged 0.7.4

Mon Oct 17 11:19:49 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Doc updates for release 0.7.4

Fri Aug 26 11:07:41 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fixes for compilation warnings (thanks to Omar Cornut)

Mon Aug 15 16:01:28 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Cleanup

Sun Aug 14 22:07:41 NZST 2005  Stephane Popinet <popinet@users.sf.net>
  * New function gts_bbox_overlaps_segment()
  
  Using the same fast routine as gts_bbox_overlaps_triangle() but with a
  degenerate triangle as input.

Wed Jul 20 16:26:02 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Removed attribute byte count check for stl2gts

Wed Jul 13 11:27:25 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Added ChangeLog target

Tue Jul 12 16:02:33 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Version set to 0.7.4 prior to release

Fri Jul  1 10:53:59 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * GtsFile now reports errors starting at column number 1

Fri Jul  1 10:53:09 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fix for comments not being fully ignored by GtsFile

Thu Jun 30 17:44:52 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fix for customised handling of parsing of GtsFile

Thu Jun 30 17:44:00 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * gts_file_new_from_string() makes a copy of the string

Wed Jun 29 13:27:34 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Small change to interface of gts_file_new_from_string()

Sun Jun 12 21:00:09 NZST 2005  Stephane Popinet <popinet@users.sf.net>
  * new gts_file_new_from_string() function

Fri Jun 10 12:26:49 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * added const qualifier to gts_object_class_from_name()

Tue Jun  7 15:52:53 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fixed bug in gts_delaunay_remove_vertex

Tue Jun  7 15:41:56 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fix for degenerate case in isotetra_bcl()
  
  It handles the degenerate case where an interpolated point
  falls directly on a mesh vertex (the previous code would generate
  edges with length zero).

Tue Jun  7 15:41:30 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Added Magnus to thanks

Tue Jun  7 15:40:46 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Fixed gts.def

Tue Jun  7 15:28:20 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Support binary STL files in stl2gts on win32

Thu May 26 15:19:54 NZST 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Added autogen.sh

Wed Mar  9 15:26:55 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * delaunay is installed

Wed Jan 19 18:41:50 NZDT 2005  Stephane Popinet <s.popinet@niwa.co.nz>
  * Updates for switch to darcs

Mon Nov 15 12:44:04 NZDT 2004  Stephane Popinet <popinet@users.sourceforge.net>
  * Bug fixes for gts_delaunay_add_constraint (s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-12)
  s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-12
  Keywords: 
  
  This function was returning destroyed conflicting constraints.
  Improper handling of special cases in remove_intersected_edge was
  causing failure in some cases.
  

Tue Nov  9 18:20:14 NZDT 2004  Stephane Popinet <popinet@users.sourceforge.net>
  * closest face for glib >= 2.4 did not use planar distance (s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-11)
  s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-11
  Keywords: 
  
  This was fixed for the glib 1.2 version a while ago.
  

Thu Nov  4 15:12:42 NZDT 2004  Stephane Popinet <popinet@users.sourceforge.net>
  * New function gts_triangle_interpolate_height() (s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-10)
  s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-10
  Keywords: 
  
  

Thu Oct 28 18:46:25 NZDT 2004  Stephane Popinet <popinet@users.sourceforge.net>
  * Added fast triangle-box intersection test (s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-9)
  s.popinet@niwa.co.nz--2004/gts--mainline--0.7--patch-9
  Keywords: 
  

sha2.h
/*************************************************************

    This program is a C++ implementation of the Secure Hash Algorithm (SHA)
    that handles the variations from the original 160 bit to 224, 256, 384
    and 512 bit.  The program is intended to be platform independant and
    has been tested on little-endian (Intel) and big-endian (Sun) machines.

    This program is based on a C version written by Aaron D. Gifford
    (as of 11/22/2004 his code could be found at http://www.adg.us/computers/sha.html).
    Attempts to contact him were unsuccessful.  I greatly condensed his version
    and shared as much code and data as I could think of.  I also inlined
    a lot of code that were macros in his version.  My version detects
    endian-ness automatically and adjusts itself accordingly.  This program
    has been tested with Visual C++ versions 6/7 and Dev-C++ on Windows, 
    g++ on Linux and CC on Solaris (g++ on Solaris gave a bus error).

    While I did make half-hearted attempts to optimize as I went along
    (testing on Wintel), any serious attempt at fast implementation is
    probably going to need to make use of in-lined assembly which is not
    very portable.

    The goal of this implementation is ease of use.  As much as possible
    I tried to hide implementation details while making it trivial to change
    the size of the hash and get the results.  The string and charactar
    array value of the hash is supplied as human-readable hex; the raw value
    can also be obtained.

    If you use this implementation somewhere I would like to be credited
    with my work (a link to my page below is fine).  I add no license
    restriction beyond any that is made by the original author.  This
    code comes with no warrenty expressed or implied, use at your own
    risk!

    Keith Oxenrider
    koxenrider[at]sol[dash]biotech[dot]com
    The latest version of this code should be available via the page
    sol-biotech.com/code.

*************************************************************/




#ifndef __SHA2C_H__
#define __SHA2C_H__

#include <string>
#include <stdexcept>
#include <string.h>
// NOTE: You may need to define things by hand for your system: 
typedef unsigned char  sha_byte;           // Exactly 1 byte 
typedef unsigned int sha_word32;           // Exactly 4 bytes 
#ifdef WIN32
    #include <windows.h>
    typedef ULONG64 sha_word64;            // 8-bytes (64-bits) 
#else
    typedef unsigned long long sha_word64; // 8-bytes (64-bits) 
#endif

// Digest lengths for SHA-1/224/256/384/512 
const sha_word32 SHA1_DIGESTC_LENGTH = 20;
const sha_word32 SHA1_DIGESTC_STRING_LENGTH   = (SHA1_DIGESTC_LENGTH   * 2 + 1);
const sha_word32 SHA224_DIGESTC_LENGTH = 28;
const sha_word32 SHA224_DIGESTC_STRING_LENGTH = (SHA224_DIGESTC_LENGTH * 2 + 1);
const sha_word32 SHA256_DIGESTC_LENGTH = 32;
const sha_word32 SHA256_DIGESTC_STRING_LENGTH = (SHA256_DIGESTC_LENGTH * 2 + 1);
const sha_word32 SHA384_DIGESTC_LENGTH = 48;
const sha_word32 SHA384_DIGESTC_STRING_LENGTH = (SHA384_DIGESTC_LENGTH * 2 + 1);
const sha_word32 SHA512_DIGESTC_LENGTH = 64;
const sha_word32 SHA512_DIGESTC_STRING_LENGTH = (SHA512_DIGESTC_LENGTH * 2 + 1);

class sha2{
public:
    enum SHA_TYPE{
        enuSHA_NONE,
        enuSHA1,
        enuSHA160 = enuSHA1,
        enuSHA224,
        enuSHA256,
        enuSHA384,
        enuSHA512,
        enuSHA_LAST //for easier looping during testing
    };

    sha2(){
        m_Type = enuSHA_NONE;
        m_boolIsBigEndian = true;
        m_boolEnded = false;

        //run-time check for endian-ness
        unsigned int test = 1;
        unsigned char *ptr = (unsigned char *)&test;
        if (ptr[0]) m_boolIsBigEndian = false;

        //these checks here because I wasn't able to figure out how to 
        //check at compile time
        if (sizeof(sha_byte) != 1) throw std::runtime_error("sha_byte != 1!");
        if (sizeof(sha_word32) != 4) throw std::runtime_error("sha_word32 != 4!");
        if (sizeof(sha_word64) != 8) throw std::runtime_error("sha_word64 != 8!");

        memset(m_chrRawHash, 0, SHA512_DIGESTC_LENGTH);
        memset(m_chrHexHash, 0, SHA512_DIGESTC_STRING_LENGTH);
        memset(m_digest, 0, SHA512_DIGESTC_LENGTH);
    };

    SHA_TYPE GetEnumType(){return m_Type;};
    bool IsBigEndian(){return m_boolIsBigEndian;};
    const char * GetTypeString(){
        switch (m_Type){
            case sha2::enuSHA1   : return "SHA160";
            case sha2::enuSHA224 : return "SHA224";
            case sha2::enuSHA256 : return "SHA256";
            case sha2::enuSHA384 : return "SHA384";
            case sha2::enuSHA512 : return "SHA512";
            default : return "Unknown!";
        }
    };

//call these three in order if you want to load chunk-by-chunk...
    void Init(SHA_TYPE type);
    //these two throw a std::runtime_error if the type is not defined
    void Update(const sha_byte *data, size_t len);//call as many times as needed
    void End();

//or call this one if you only have one chunk of data
    const std::string &GetHash(SHA_TYPE type, const sha_byte* data, size_t len);

//call one of these routines to access the hash
    //these throw a std::runtime_error if End has not been called
    const char *HexHash();//NULL terminated
    const std::string &StringHash();
    const char *RawHash(int &length);//NO NULL termination! size stored in 'length'


private:
    SHA_TYPE m_Type;
    std::string m_strHash;
    bool m_boolEnded, m_boolIsBigEndian;
    char m_chrRawHash[SHA512_DIGESTC_LENGTH], m_chrHexHash[SHA512_DIGESTC_STRING_LENGTH];
    sha_byte m_digest[SHA512_DIGESTC_LENGTH];

//these are common buffers for maintaining the hash
    struct SHA_CTX{
        sha_byte   state[sizeof(sha_word64) * 8];//maximum size
        sha_word64 bitcount[2];//sha1, 224 and 256 only use the first entry
        sha_byte   buffer[128];
    }ctx;

    
//** INTERNAL FUNCTION PROTOTYPES ************************************
    void SHA256_Internal_Last(bool isSha1 = false);
    void SHA512_Internal_Last();

    void SHA1_Internal_Transform(const sha_word32 *data);
    void SHA256_Internal_Transform(const sha_word32* data);
    void SHA512_Internal_Transform(const sha_word64*);

    void SHA32bit_Update(const sha_byte *data, size_t len, bool isSha1=false);
    void SHA64bit_Update(const sha_byte *data, size_t len);

//macro replacements
    inline void MEMSET_BZERO(void *p, size_t l){memset(p, 0, l);};
    inline void MEMCPY_BCOPY(void *d,const void *s, size_t l) {memcpy(d, s, l);};

    //For incrementally adding the unsigned 64-bit integer n to the
    //unsigned 128-bit integer (represented using a two-element array of
    //64-bit words):
    inline void ADDINC128(sha_word64 *w, sha_word32 n)  {
        w[0] += (sha_word64)(n);
        if (w[0] < (n)) w[1]++;
    }

    // Shift-right (used in SHA-256, SHA-384, and SHA-512): 
    inline sha_word32 SHR(sha_word32 b,sha_word32 x){return (x >> b);};
    inline sha_word64 SHR(sha_word64 b,sha_word64 x){return (x >> b);};
    // 32-bit Rotate-right (used in SHA-256): 
    inline sha_word32 ROTR32(sha_word32 b,sha_word32 x){return ((x >> b) | (x << (32 - b)));};
    // 64-bit Rotate-right (used in SHA-384 and SHA-512): 
    inline sha_word64 ROTR64(sha_word64 b,sha_word64 x){return ((x >> b) | (x << (64 - b)));};
    // 32-bit Rotate-left (used in SHA-1): 
    inline sha_word32 ROTL32(sha_word32 b,sha_word32 x){return ((x << b) | (x >> (32 - b)));};

    // Two logical functions used in SHA-1, SHA-254, SHA-256, SHA-384, and SHA-512: 
    inline sha_word32 Ch(sha_word32 x,sha_word32 y,sha_word32 z){return ((x & y) ^ ((~x) & z));};
    inline sha_word64 Ch(sha_word64 x,sha_word64 y,sha_word64 z){return ((x & y) ^ ((~x) & z));};
    inline sha_word32 Maj(sha_word32 x,sha_word32 y,sha_word32 z){return ((x & y) ^ (x & z) ^ (y & z));};
    inline sha_word64 Maj(sha_word64 x,sha_word64 y,sha_word64 z){return ((x & y) ^ (x & z) ^ (y & z));};

    // Function used in SHA-1: 
    inline sha_word32 Parity(sha_word32 x,sha_word32 y,sha_word32 z){return (x ^ y ^ z);};

// Four logical functions used in SHA-256: 
    inline sha_word32 Sigma0_256(sha_word32 x){return (ROTR32(2, x) ^ ROTR32(13, x) ^ ROTR32(22, x));};
    inline sha_word32 Sigma1_256(sha_word32 x){return (ROTR32(6, x) ^ ROTR32(11, x) ^ ROTR32(25, x));};
    inline sha_word32 sigma0_256(sha_word32 x){return (ROTR32(7, x) ^ ROTR32(18, x) ^ SHR(   3 , x));};
    inline sha_word32 sigma1_256(sha_word32 x){return (ROTR32(17,x) ^ ROTR32(19, x) ^ SHR(   10, x));};

// Four of six logical functions used in SHA-384 and SHA-512: 
    inline sha_word64 Sigma0_512(sha_word64 x){return (ROTR64(28, x) ^ ROTR64(34, x) ^ ROTR64(39, x));};
    inline sha_word64 Sigma1_512(sha_word64 x){return (ROTR64(14, x) ^ ROTR64(18, x) ^ ROTR64(41, x));};
    inline sha_word64 sigma0_512(sha_word64 x){return (ROTR64( 1, x) ^ ROTR64( 8, x) ^ SHR(    7, x));};
    inline sha_word64 sigma1_512(sha_word64 x){return (ROTR64(19, x) ^ ROTR64(61, x) ^ SHR(    6, x));};

    inline void REVERSE32(sha_word32 w, sha_word32 &x)  {
        w = (w >> 16) | (w << 16);
        x = ((w & 0xff00ff00UL) >> 8) | ((w & 0x00ff00ffUL) << 8);
    }
    #ifdef _VC6
        inline void REVERSE64(sha_word64 w, sha_word64 &x)  {
            w = (w >> 32) | (w << 32);
            w =   ((w & 0xff00ff00ff00ff00ui64) >> 8) |
                  ((w & 0x00ff00ff00ff00ffui64) << 8);
            (x) = ((w & 0xffff0000ffff0000ui64) >> 16) |
                  ((w & 0x0000ffff0000ffffui64) << 16);
        }
    #else
        inline void REVERSE64(sha_word64 w, sha_word64 &x)  {
            w = (w >> 32) | (w << 32);
            w =   ((w & 0xff00ff00ff00ff00ULL) >> 8) |
                  ((w & 0x00ff00ff00ff00ffULL) << 8);
            (x) = ((w & 0xffff0000ffff0000ULL) >> 16) |
                  ((w & 0x0000ffff0000ffffULL) << 16);
        }
    #endif

};//end class sha2
#endif // __SHA2C_H__ 
